orca-core/src/main/groovy/com/netflix/spinnaker/orca/batch/adapters/TaskTasklet.groovy

```diff
@@ -77,8 +77,8 @@ class TaskTasklet implements Tasklet {
         setStopStatus(chunkContext, ExitStatus.STOPPED, ExecutionStatus.CANCELED)
         contribution.exitStatus = ExitStatus.STOPPED
         return cancel(stage)
-      } else if (task.status.complete || task.status.halt) {
-        // no-op
+      } else if (!OptionalStageSupport.isOptional(stage) && (task.status.complete || task.status.halt)) {
```

Comment:
The issue is that a subsequent restart would see the tasks as complete even if the overall stage became optional.